p2-maven-plugin/src/main/java/org/eclipse/tycho/p2maven/MavenProjectDependencyProcessor.java

```diff
@@ -140,12 +157,74 @@ public Collection<IInstallableUnit> computeProjectDependencies(MavenProject proj
 		if (projectUnits.isEmpty()) {
 			return Collections.emptyList();
 		}
-		IQueryResult<IInstallableUnit> result = slicer.computeDependencies(projectUnits, avaiableIUs);
-		Set<IInstallableUnit> resolved = result.toSet();
+		Set<IInstallableUnit> resolved = slicer.computeDependencies(projectUnits, avaiableIUs).toSet();
 		resolved.removeAll(projectUnits);
+		// TODO maybe contribute this part to P2 Slicer?
+		if (!resolved.isEmpty()) {
+			// we now need to attach all fragments to resolved units
+			List<IInstallableUnit> dependentFragments = new ArrayList<IInstallableUnit>();
+			List<IRequiredCapability> hosts = projectUnits.stream()
+					.flatMap(iu -> getFragmentHostRequirement(iu).stream()).collect(Collectors.toList());
+			CollectionResult<IInstallableUnit> collectionResult;
+			if (hosts.isEmpty()) {
+				collectionResult = new CollectionResult<IInstallableUnit>(resolved);
+			} else {
+				// we must filter out our host here, as otherwise we pull in other fragments as
+				// dependencies and produce a cycle...
+				List<IInstallableUnit> filtered = resolved.stream().filter(iu -> {
+					for (IRequiredCapability host : hosts) {
+						if (host.isMatch(iu)) {
+							return false;
+						}
+					}
+					return true;
+				}).collect(Collectors.toList());
+				collectionResult = new CollectionResult<IInstallableUnit>(filtered);
+			}
+			for (IInstallableUnit unit : avaiableIUs.query(QueryUtil.ALL_UNITS, null)) {
+				if (isFragment(unit) && !projectUnits.contains(unit)) {
+					Set<IInstallableUnit> fragmentsResult = slicer
+							.computeDependencies(Collections.singleton(unit), collectionResult).toSet();
+					if (fragmentsResult.size() > 1) {
+						// at least one of the resolved items depend on the fragment
+						dependentFragments.add(unit);
+					}
+				}
+			}
+			resolved.addAll(dependentFragments);
+		}
 		return resolved;
 	}
 
+	private static boolean isFragment(IInstallableUnit installableUnit) {
+		return getFragmentCapability(installableUnit).isPresent();
+	}
+
+	private static Optional<IProvidedCapability> getFragmentCapability(IInstallableUnit installableUnit) {
+		for (IProvidedCapability capability : installableUnit.getProvidedCapabilities()) {
+			String nameSpace = capability.getNamespace();
```

Comment:
I wonder if it would in general lead to a performance improvement for P2 (and Tycho) if capabilities/requirements could be retrieved per namespace and would be stored internally per namespace, just like it is now done in Equinox for bundle requirements/capabilities. 
When switching to the Equinox-resolver Tycho first suffered from a massive increase of the initial target resolution just because of that: https://github.com/eclipse/tycho/issues/47

Reply:
The problem scope is much smaller here so it should not be an issue, nerveless a lookup by name could be useful I think, it is just a bit out of scope of this changes, as the first priority would be correctness and could later be optimized for speed :-)

Reply:
> The problem scope is much smaller here so it should not be an issue, nerveless a lookup by name could be useful I think, it is just a bit out of scope of this changes, as the first priority would be correctness and could later be optimized for speed :-) SO if you like to propose such a change for P2 it would be welcome.

Yes this is definitely related to this change and just came into my mind when locking at the change.
I'm going to create an issue, however I did not yet check the P2 sources if this is in general useful.